### PR TITLE
stackruntime: Support referencing Local Values within the `tfstacks.hcl` language

### PR DESCRIPTION
**Pull request by mutahhir**
Local Values are made available in the .tfstack.hcl language. The existing codebase already had support for local values, it was incomplete was missing the machinery required to reference local values within the `tfstacks.hcl` files. 

## Target Release

1.8.x

### CODE REVIEW

File: internal/stacks/stackruntime/internal/stackeval/local_value.go
Comment (apparentlymart) [inline internal/stacks/stackruntime/internal/stackeval/local_value.go:92]: Indeed, DynamicVal is a reasonable fallback here, for the reason you stated. We try to be more specific where possible, but local values prioritize convenience (not having to explicitly specify a type) over type checking -- because they are always local within a particular stack anyway, so the consequences of getting it "wrong" are contained.

A consequence of that convenience is that we don't have type information when the expression is too invalid to evaluate, but cty will still propagate this correctly and may be able to do more inference downstream if this value is used in a place that implies a type.

Comment (mutahhir) [reply internal/stacks/stackruntime/internal/stackeval/local_value.go:92]: Thank you for taking a look and explaining it!

File: internal/stacks/stackruntime/internal/stackeval/local_value_config.go
Comment (apparentlymart) [inline internal/stacks/stackruntime/internal/stackeval/local_value_config.go:17]: This comment is referring to "variable" blocks rather than "locals" blocks.

File: internal/stacks/stackruntime/internal/stackeval/local_value_config.go
Comment (apparentlymart) [inline internal/stacks/stackruntime/internal/stackeval/local_value_config.go:64]: What does this string represent? 🤔

File: internal/stacks/stackruntime/internal/stackeval/local_value_test.go
Comment (apparentlymart) [inline internal/stacks/stackruntime/internal/stackeval/local_value_test.go:83]: Let's remove unneeded code before merging, just so it's not confusing to a future maintainer.

File: internal/stacks/stackruntime/internal/stackeval/planning_test.go
Comment (apparentlymart) [inline internal/stacks/stackruntime/internal/stackeval/planning_test.go:925]: Is this comment accurate for this test? It seems to be talking about input variables, so I'm wondering if this comment belonged to another test that you derived this one from.

**Review by apparentlymart — APPROVED** [review]
This looks good to me!

My comments inline are mostly just little details that I don't think affect the end-user-visible behavior, but might make things clearer for the next maintainer.

File: internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl
Comment (mutahhir) [inline internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl:9]: Add the following tests:

- Add a local that references a data source (implement a dummy provider, that exposes a data source).
- Add a local that references a data source which returns an unknown value.

File: internal/stacks/stackruntime/internal/stackeval/local_value.go
Comment (mutahhir) [reply internal/stacks/stackruntime/internal/stackeval/local_value.go]: Thank you for the insight! I've removed the code now

File: internal/stacks/stackruntime/internal/stackeval/local_value_config.go
Comment (mutahhir) [reply internal/stacks/stackruntime/internal/stackeval/local_value_config.go]: This was for testing if we encounter this path somewhere, but it didn't seem like it showed up, so I've replaced it with DynamicVal seeing as how the type was stored for referencing later. I wonder if that's correct as well. Would love to learn more about whether this interface implementation is required for locals and how it's used in the scheme of things.

File: internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl
Comment (mutahhir) [reply internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl:9]: @alisdair I finally got a test running for the first part of this, which basically tests whether locals work fine with data sources. Do you recall what the 'unknown value' return by the data source means here?

File: internal/stacks/stackruntime/internal/stackeval/local_value_config.go
Comment (alisdair) [reply internal/stacks/stackruntime/internal/stackeval/local_value_config.go]: I believe that `ExprReferenceValue` is called on descendants of `StackConfig` during validation. Some configuration constructs are able to give more informative results than just an unknown value of unknown type (e.g. embedded stack calls). It seems valid to me to return that in this case since we won't always know anything about the value or its type here.

File: internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl
Comment (alisdair) [reply internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl:9]: I think the idea with that test was to ensure that unknowns are propagated correctly, and the local value can return an unknown. The "data source" part isn't particularly helpful, and I think we must have meant "a resource planned with unknown values"—more often that'll be a managed resource with an attribute like `"id"` or the like.

File: internal/stacks/stackruntime/internal/stackeval/local_value_config.go
Comment (apparentlymart) [reply internal/stacks/stackruntime/internal/stackeval/local_value_config.go]: If I'm understanding the rest of `local_value_config.go` correctly, I think a plausible implementation of `ExprReferenceValue` here would be to call `v.Validate` value and then return its first return value, discarding the diagnostics.

That would then return whatever is our best approximation of the local value based on the information we have during validation. In the best case where the local value is defined as a literal value it would be that value. If it were defined as a reference to something else we don't know during validation then it would instead be an unknown value placeholder for that thing.

Comment (apparentlymart) [reply internal/stacks/stackruntime/internal/stackeval/local_value_config.go]: (`validateValueInner` should probably also take a `phase EvalPhase` argument and pass it in to the expression evaluator, rather than hard-coding `ValidatePhase`, because these config eval functions also get called during `PlanPhase` as a static check before we do the dynamic evaluation. It might make things clearer to rename `ValidateValue` and `validateValueInner` to be `CheckValue` and `checkValueInner` to avoid the implication that they are used only during the validate phase; I suspect that naming was a remnant of some early work when it wasn't clear to me yet exactly how we'd handle plan-time static checks.)

File: internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl
Comment (mutahhir) [reply internal/stacks/stackruntime/internal/stackeval/testdata/sourcebundle/local_value/basics/local-value-basics.tfstack.hcl:9]: I think the last test added is broad enough to cover both these cases then. It takes a data source, and the component outputs that value which is then stored within the local and then passed as input to another component. Should that cover the 'unknown' concept enough?

**Review by DanielMSchmidt — APPROVED** [review]
Looks great to me, I feel like the test cases cover enough ground 👍

**Pull request merged**
